change_analyzer/sequences_diff.py

```diff
@@ -269,6 +336,149 @@ def write_to_report(self):
+            actual_page_xml = series[f'{SequenceRecorder.COL_PAGE_SOURCE_AFTER}_actual']
+        else:
+            # We are not in the first step and we need to compare the Before page sources
+            expected_page_xml = series[f'{SequenceRecorder.COL_PAGE_SOURCE_BEFORE}_expected']
+            actual_page_xml = series[f'{SequenceRecorder.COL_PAGE_SOURCE_BEFORE}_actual']
+
+        expected_xml_encoding = self.get_xml_encoding(expected_page_xml)
+        actual_xml_encoding = self.get_xml_encoding(actual_page_xml)
+
+        # Encode to fix error ValueError: Unicode strings with encoding declaration are not supported.
+        # Please use bytes input or XML fragments without declaration.
+        expected_page_xml_encoded = expected_page_xml.encode(expected_xml_encoding)
+        actual_page_xml_encoded = actual_page_xml.encode(actual_xml_encoding)
+
+        diff_results = self.extract_diffs_as_dict(expected_page_xml_encoded, actual_page_xml_encoded)
+
+        for k, v in diff_results['actual'].items():
+            expected_element = diff_results['expected'][k]
+            actual_element = diff_results['actual'][k]
+            expected_x = expected_element['x']
+            actual_x = actual_element['x']
+            expected_y = expected_element['y']
+            actual_y = actual_element['y']
+            expected_width = expected_element['width']
+            actual_width = actual_element['width']
+            expected_height = expected_element['height']
+            actual_height = actual_element['height']
+            expected_box_coordinates = self._get_box_coordinates(expected_x, expected_y,
+                                                                 expected_width, expected_height)
+            actual_box_coordinates = self._get_box_coordinates(actual_x, actual_y,
+                                                               actual_width, actual_height)
+            info = f'Actual ({actual_box_coordinates}) vs expected ({expected_box_coordinates})'
+            info_dict = {
+                'actual_element': actual_element,
+                'expected_element': expected_element,
+                'actual_element_boundaries': actual_box_coordinates,
+                'expected_element_boundaries': expected_box_coordinates,
+                'info': info
+            }
+            step_diff_info.append(info_dict)
+
+        self.diff_info.append(step_diff_info)
+
+    @staticmethod
+    def _get_box_coordinates(x: str, y: str, w: str, h: str) -> List[Tuple[int, int]]:
+        """Get boundary coordinates based on x, y, width and height"""
+        x = int(x)
+        y = int(y)
+        w = int(w)
+        h = int(h)
+        return [(x, y), (x+w, y), (x+w, y+h), (x, y+h)]
+
+    def _save_dataframe_to_csv(self) -> None:
+        """Save merged dataframe to the csv"""
+        csv_file = self.report_file.split(".")[0] + ".csv"
+        self.df_merged.to_csv(csv_file, index=False)
+

```

Comment:
500 lines is a bit too big for a single class. Probably would be good to extract PageSourceDiff class.